### PR TITLE
feat: add generic-sqlalchemy executor to support most of databases

### LINKED ISSUE #498: Separate out optional package installation to speed up demo installation
Motivation: Querybook installs many dependencies but a typical user/company would not use most of them. We should refactor base.txt to only install sqlalchemy for support sqlite querying and provide every other engine installation as on demand.

Requirement:
- Refactor requirements.txt to separate out must-required dependencies vs custom dependencies
- Figure out a simple way for users to install these dependencies on-demand without using the plugins
- Document the installation steps for every type of query engine querybook can support and instructions on how

### PR DESCRIPTION
**Pull request by czgu**
Closes #498. We can take out the optional hive/druid/snowflake dependencies on the v3 update later since it is an incompatible change.

This adds the generic-sqlalchemy executor which would work for any sqlalchemy compatible databases. It also auto checks with sqlalchemy to see if the library required is available before showing it in the admin UI. For example, if you install the sqlalchemy-redshift package, then you would see this
image

I have also updated the doc to explain how to use this executor and covered our current support level for every database.
image
image

### CODE REVIEW

File: docs_website/docs/setup_guide/connect_to_query_engines.md [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+| Query Progress                 | x          | ?                 | ✓                  |
+| Query Logs                     | x          | ?                 | ✓                  |
+| Query Metadata                 | x          | ?                 | ✓                  |
+| Cancel Query                   | x          | ?                 | ✓                  |
+| User Authentication            | x          | x                 | ✓                  |
+| Syntax Error Parsing           | x          | ?                 | ✓                  |
+| Service discovery              | x          | x                 | ✓                  |
+| Language Specific Autocomplete | x          | x                 | ✓                  |
+
+Tier 1 engines do not mean they can be used in production everywhere since different companies/org require different kinds of integrations. However, tier 1 databases provide an excellent foundation to extend additional functionalities. Use them as a reference or subclass them via the [query engine plugin](../integrations/add_query_engine.md).
```
Comment (meowcodes) [inline docs_website/docs/setup_guide/connect_to_query_engines.md:24]: Tier 1 does not mean engines can be

File: docs_website/docs/setup_guide/connect_to_query_engines.md [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+
+If you have tried any of the tier 3 databases and confirmed it works, please update this doc to let others know.
+
+## Query Engine Support
+
+Querybook only supports a few of the Tier 2/1 databases by default. When Querybook is launched, it checks with SqlAlchemy to see if any of the databases below are available. If so, the query engine would be automatically available to set up in the Admin UI. Please see the [step by step guide](#step-by-step-guide) below to see an working example.
```
Comment (meowcodes) [inline docs_website/docs/setup_guide/connect_to_query_engines.md:30]: Tier 2 & 1

File: docs_website/docs/setup_guide/connect_to_query_engines.md [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+6. Visit [https://localhost:10001/admin/query_engine/](https://localhost:10001/admin/query_engine/) and create a new query engine. Put `redshift` as the language and `generic-sqlalchemy` as the executor. In the `Executor Params`, put the connection string (as specified by SqlAlchemy) in the `Connection_string` field.
+7. Go to [https://localhost:10001/admin/environment/1/](https://localhost:10001/admin/environment/1/) and add the Redshift engine under the demo_environment.
+8. Now you can run queries against the new Redshift engine in [https://localhost:10001/demo_environment/adhoc/](https://localhost:10001/demo_environment/adhoc/).
+9. To include table metadata and autocompletion, you would need to add a metastore. Visit [https://localhost:10001/admin/metastore/](https://localhost:10001/admin/metastore/) and create a new metastore. Use SqlAlchemyMetastoreLoader with the exac connection string used for the query engine. Click on `Save` -> `CREATE SCHEDULE` -> `Create Task`. Now click on `Run Task` to sync. You can view the progress in the `History` tab. Wait until it is completed (Should be done in seconds if the number of tables is small).
```
Comment (meowcodes) [inline docs_website/docs/setup_guide/connect_to_query_engines.md:65]: Use SqlAlchemyMetastoreLoader with the **exact** connection

File: docs_website/docs/setup_guide/connect_to_query_engines.md [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+10. Go to your query engine page on [https://localhost:10001/admin/query_engine/](https://localhost:10001/admin/query_engine/), in the Metastore field, choose the metastore you just created and click `Save`.
+11. Visit [https://localhost:10001/demo_environment/adhoc/](https://localhost:10001/demo_environment/adhoc/) again and the auto complete feature should be available. You can also view all tables by clicking on the `Tables` button on the left sidebar and then select the specific metastore.
```
Comment (meowcodes) [inline docs_website/docs/setup_guide/connect_to_query_engines.md:67]: and select (remove then)

File: docs_website/docs/setup_guide/connect_to_query_engines.md [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+
+If you have tried any of the tier 3 databases and confirmed it works, please update this doc to let others know.
+
+## Query Engine Support
+
+Querybook only supports a few of the Tier 2/1 databases by default. When Querybook is launched, it checks with SqlAlchemy to see if any of the databases below are available. If so, the query engine would be automatically available to set up in the Admin UI. Please see the [step by step guide](#step-by-step-guide) below to see an working example.
+
+## Step by step guide
+
+In this guide, we will through all the steps required to get Amazon Redshift to work in Querybook. Other query engines can be added similarly.
```
Comment (meowcodes) [inline docs_website/docs/setup_guide/connect_to_query_engines.md:34]: In this guide, we will go through adding Amazon Redshift query engine to Querybook. This serves as an example to adding all query engines.

**Review by meowcodes — APPROVED** [review]

File: docs_website/docs/setup_guide/connect_to_query_engines.md (hunk 3 above)
Comment (czgu) [reply docs_website/docs/setup_guide/connect_to_query_engines.md:65]: wow good catch

**Pull request merged**